pumps/sql.go

```diff
@@ -354,3 +370,82 @@ func (c *SQLPump) WriteUptimeData(data []interface{}) {
 
 	c.log.Debug("Purged ", len(data), " records...")
 }
+
+func (c *SQLPump) buildIndexName(indexBaseName, tableName string) string {
+	return fmt.Sprintf("%s_%s", tableName, indexBaseName)
+}
+
+// ensureIndex check that all indexes for the analytics SQL table are in place
+func (c *SQLPump) ensureIndex(tableName string, background bool) error {
+	if !c.db.Migrator().HasTable(tableName) {
+		return errors.New("cannot create indexes as table doesn't exist: " + tableName)
+	}
+
+	// waitgroup to facilitate testing and track when all indexes are created
+	var wg sync.WaitGroup
+	wg.Add(len(indexes))
+	createIndexFn := func(indexBaseName, column string) error {
+		defer wg.Done()
+		indexName := c.buildIndexName(indexBaseName, tableName)
+		option := ""
+		if c.dbType == "postgres" {
+			option = "CONCURRENTLY"
+		}
+
+		columnExist := c.db.Migrator().HasColumn(&analytics.AnalyticsRecord{}, column)
+		if !columnExist {
+			return errors.New("cannot create index for non existent column " + column)
+		}
+
+		sql := fmt.Sprintf("CREATE INDEX %s IF NOT EXISTS %s ON %s (%s)", option, indexName, tableName, column)
+		err := c.db.Exec(sql).Error
+		if err != nil {
+			c.log.Errorf("error creating index %s for table %s : %s", indexName, tableName, err.Error())
+			return err
+		}
+
+		c.log.Infof("Index %s created for table %s", indexName, tableName)
+		return nil
+	}
+
+	for _, idx := range indexes {
+		indexName := tableName + idx.baseName
+
+		if !c.db.Migrator().HasIndex(tableName, indexName) {
```

Comment:
let's use early exits for better readability and to avoid nesting
```suggestion
		if c.db.Migrator().HasIndex(tableName, indexName) {
		  c.log.WithFields(logrus.Fields{
				"index": indexName,
				"table": tableName,
			}).Info("index already exists")
		continue
		}
		if background {...
```